File: 0543-diameter-of-binary-tree/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    depth = 0
    def getDepth(self, root):
        if not root:
            return 0
        left = self.getDepth(root.left);
        right = self.getDepth(root.right)
        self.depth = max(self.depth, left + right)
        return max(left, right) + 1
        
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        self.getDepth(root)
        return self.depth

File: 0543-diameter-of-binary-tree/solution.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
int max(int num1, int num2) {
    return num1 > num2 ? num1 : num2;
}
int getMaxDepth(struct TreeNode* root, int *length) {
    if (root == NULL) 
        return 0;
    int totalLeft = getMaxDepth(root->left, length);
    int totalRight = getMaxDepth(root->right, length);
    *length = max(*length, totalLeft + totalRight);
    return max(totalLeft, totalRight) + 1;
}
int diameterOfBinaryTree(struct TreeNode* root){
    int answer = 0;
    getMaxDepth(root, &answer);
    return answer;
}
